cmd/transcriber/call/live_captions.go

```diff
@@ -0,0 +1,461 @@
+					MetricName: public.MetricLiveCaptionsTranscriberBufFull,
+				}, false); err != nil {
+					slog.Error("processLiveCaptionsForTrack: error sending wsEvMetric MetricTranscriberBufFull",
+						slog.String("err", err.Error()),
+						slog.String("trackID", ctx.trackID))
+				}
+				close(transcribedCh)
+			}
+
+			// While audio is being transcribed, we need to cut down the window if it's > maxWindowSize.
+			window, prevTranscribedPos = cutWindowToSize(ctx.trackID, window, segments, prevTranscribedPos)
+			prevWindowLen = len(window)
+
+			// Use a for loop and a select so that we can drop ticks waiting for the transcriber.
+		waitForTranscription:
+			for {
+				select {
+				case <-ticker.C:
+					slog.Debug("processLiveCaptionsForTrack: dropped a tick waiting for the transcriber",
+						slog.String("trackID", ctx.trackID))
+				case text := <-transcribedCh:
+					if len(text) == 0 {
+						// Either transcribedCh was closed above (captionQueueCh full), or audio transcription failed.
+						// Note: this appears to happen when the transcriber fails to decode a block of audio.
+						// Usually the probability returned for the language is very low, which makes sense.
+						slog.Debug("processLiveCaptionsForTrack: received empty text, ignoring.")
+						break waitForTranscription
+					}
+					if err := t.client.SendWs(wsEvCaption, public.CaptionMsg{
+						SessionID:     ctx.sessionID,
+						UserID:        ctx.user.Id,
+						Text:          text,
+						NewAudioLenMs: float64(newAudioLenMs),
+					}, false); err != nil {
+						slog.Error("processLiveCaptionsForTrack: error sending ws captions",
+							slog.String("err", err.Error()),
+							slog.String("trackID", ctx.trackID))
+					}
+
+					break waitForTranscription
+				}
+			}
+		}
+	}
+}
+
+type segmentSamples struct {
+	Start   int
+	End     int
+	Silence bool
+}
+
+// convertToSegmentSamples turns the speech.Segments (in time) into segmentSamples (measured in samples)
+func convertToSegmentSamples(segments []speech.Segment, audioLen int) []segmentSamples {
+	var ret []segmentSamples
+	lastEndAtIdx := 0
+	for _, seg := range segments {
+		start := int(seg.SpeechStartAt * trackOutAudioRate)
+		end := int(seg.SpeechEndAt * trackOutAudioRate)
+		ret = append(ret, segmentSamples{
+			Start:   lastEndAtIdx,
+			End:     start,
+			Silence: true,
+		})
+		ret = append(ret, segmentSamples{
+			Start:   start,
+			End:     end,
+			Silence: false,
+		})
+		lastEndAtIdx = end
+	}
+
+	if lastEndAtIdx < audioLen {
+		ret = append(ret, segmentSamples{
+			Start:   lastEndAtIdx,
+			End:     audioLen,
+			Silence: true,
+		})
+	}
+
+	return ret
+}
+
+// removeShortSpeeches removes small sections of speech because either they are not actual words,
+// or the transcriber will have trouble with such a short amount.
+func removeShortSpeeches(segments []segmentSamples) {
+	for i, seg := range segments {
+		if !seg.Silence && (seg.End-seg.Start) < minSpeechLengthSamples {
```

Comment:
Not too convinced by this. You can say a lot in one second. I'd rather have artifacts than data loss if I could choose.

Reply:
We do need some smoothing here, or else you get more hallucinations. One of the Silero scientists states that most meaningful speech is longer than 250ms (https://thegradient.pub/one-voice-detector-to-rule-them-all/) so I did some experiments around that number. I tended to see more hallucinations when the minSpeechLength was 150ms, but fewer at 210 and 240, then I couldn't tell a huge difference above that. So I think 210ms (actually 330 bc of padding) seems like a good barrier. This means that if we have [`silence` (>=150ms), `speech` (<=210ms + 120ms padding), `silence` (>=150ms)], we can safely remove the speech.
So yes, we're accepting a bit of data loss, but the result is fewer hallucinations and garbage output. I think that's a worthwhile tradeoff considering the type of speech we'll be losing at that level will be short "um" and "ah"s. Longer "um"s and "ah"s are still detected.

Reply:
Thanks for the thorough explanation. ~300ms sounds like a much more reasonable threshold.